lib/internal/fs/glob.js

```diff
@@ -0,0 +1,202 @@
+  ArrayPrototypePop,
+  ArrayPrototypePush,
+  SafeMap,
+  SafeSet,
+} = primordials;
+
+let minimatch;
+function lazyMinimatch() {
+  minimatch ??= require('internal/deps/minimatch/index');
+  return minimatch;
+}
+
+function testPattern(pattern, path) {
+  if (pattern === lazyMinimatch().GLOBSTAR) {
+    return true;
+  }
+  if (typeof pattern === 'string') {
+    return true;
+  }
+  if (typeof pattern.test === 'function') {
+    return pattern.test(path);
+  }
+}
+
+class Cache {
+  #caches = new SafeMap();
+  #statsCache = new SafeMap();
+  #readdirCache = new SafeMap();
+
+  stats(path) {
+    if (this.#statsCache.has(path)) {
+      return this.#statsCache.get(path);
+    }
+    let val;
+    try {
+      val = lstatSync(path);
+    } catch {
+      val = null;
+    }
+    this.#statsCache.set(path, val);
+    return val;
+  }
+  readdir(path) {
+    if (this.#readdirCache.has(path)) {
+      return this.#readdirCache.get(path);
+    }
+    let val;
+    try {
+      val = readdirSync(path, { withFileTypes: true });
+      ArrayPrototypeForEach(val, (dirent) => this.#statsCache.set(join(path, dirent.name), dirent));
+    } catch {
+      val = [];
+    }
+    this.#readdirCache.set(path, val);
+    return val;
+  }
+
+  seen(pattern, index, path) {
+    return this.#caches.get(path)?.get(pattern)?.has(index);
+  }
+  add(pattern, index, path) {
+    if (!this.#caches.has(path)) {
+      this.#caches.set(path, new SafeMap([[pattern, new SafeSet([index])]]));
+    } else if (!this.#caches.get(path)?.has(pattern)) {
+      this.#caches.get(path)?.set(pattern, new SafeSet([index]));
+    } else {
+      this.#caches.get(path)?.get(pattern)?.add(index);
+    }
+  }
+
+}
+
+function glob(patterns, options = kEmptyObject) {
+  validateObject(options, 'options');
+  const root = options.cwd ?? '.';
+  const { exclude } = options;
+  if (exclude != null) {
+    validateFunction(exclude, 'options.exclude');
+  }
+
+  const { Minimatch, GLOBSTAR } = lazyMinimatch();
+  const results = new SafeSet();
+  const matchers = ArrayPrototypeMap(patterns, (pattern) => new Minimatch(pattern));
+  const queue = ArrayPrototypeFlatMap(matchers, (matcher) => {
+    return ArrayPrototypeMap(matcher.set,
+                             (pattern) => ({ __proto__: null, pattern, index: 0, path: '.', followSymlinks: true }));
+  });
+  const cache = new Cache(matchers);
+
+  while (queue.length > 0) {
+    const { pattern, index: currentIndex, path, followSymlinks } = ArrayPrototypePop(queue);
+    if (cache.seen(pattern, currentIndex, path)) {
+      continue;
+    }
+    cache.add(pattern, currentIndex, path);
+
+    const currentPattern = pattern[currentIndex];
+    const index = currentIndex + 1;
+    const isLast = pattern.length === index || (pattern.length === index + 1 && pattern[index] === '');
+
+    if (currentPattern === '') {
+      // Absolute path
+      ArrayPrototypePush(queue, { __proto__: null, pattern, index, path: '/', followSymlinks });
```

Comment:
A pattern of `C:` at index 0 on windows should also be considered an absolute path.

UNC paths are completely different.  They look like `['', '', string, string, ...]`, but starting to match at `/` will not work, since that resolves to the drive letter on Windows.

Reply:
You can also get a pattern portion of `''` if the pattern _ends_ with a `/`, in which case it should be treated the same as `.`